lib/Targets/AIEVecToCpp/TranslateAIEVecToCpp.cpp

```diff
@@ -2598,6 +2608,85 @@ static LogicalResult printOperation(CppEmitter &emitter,
   return success();
 }
 
+static std::string printConversionTo512bit(CppEmitter &emitter, Value v) {
+  std::string vName = emitter.getOrCreateName(v).str();
+  auto vTy = cast<VectorType>(v.getType());
+  auto vShape = vTy.getShape();
+  int64_t elemBitWidth = vTy.getElementTypeBitWidth();
+  int64_t numElems = std::accumulate(vShape.begin(), vShape.end(), 1,
+                                     std::multiplies<int64_t>());
+  int64_t vBitWidth = numElems * elemBitWidth;
+  if (vBitWidth >= 512)
+    return vName;
+
+  int64_t newNumElems = 512 / elemBitWidth;
+
+  std::string vNewName = emitter.getNewName();
+  raw_indented_ostream &os = emitter.ostream();
+  auto newVecTy = VectorType::get({512 / elemBitWidth}, vTy.getElementType());
+  auto newTyName = *(
+      emitter.genCppTypeName(newVecTy, /*stdintType=*/false, /*isAcc=*/false));
+  auto oldTyName =
+      *(emitter.genCppTypeName(vTy, /*stdintType=*/false, /*isAcc=*/false));
+
+  os << newTyName << " " << vNewName << " = concat(";
+  if (newNumElems / numElems == 4) {
+    os << "concat(" << vName << ", undef_" << oldTyName << "())";
+    oldTyName = *(emitter.genCppTypeName(
+        VectorType::get({256 / elemBitWidth}, vTy.getElementType())));
+  } else {
+    os << vName;
+  }
+  os << ", undef_" << oldTyName << "());\n";
+  return vNewName;
+}
+
+static LogicalResult printOperation(CppEmitter &emitter,
+                                    aievec::MatMulOp matmulOp) {
+  auto lhs = matmulOp.getLhs();
+  auto rhs = matmulOp.getRhs();
+  auto acc = matmulOp.getAcc();
+
+  // The sources should have already been emitted
+  if (!emitter.hasValueInScope(lhs) || !emitter.hasValueInScope(rhs) ||
+      !emitter.hasValueInScope(acc))
+    return failure();
+
+  auto lhsName = printConversionTo512bit(emitter, lhs);
+  auto rhsName = printConversionTo512bit(emitter, rhs);
+
+  raw_indented_ostream &os = emitter.ostream();
+
+  // If acc is not an accumulator, load it into one:
+  StringRef accName = emitter.getOrCreateName(acc);
+  std::string accTyName = *(emitter.genCppTypeName(acc.getType()));
+  auto accTyNameOpt = emitter.genCppTypeName(
+      acc.getType(), /*stdintType=*/false, /*isAcc=*/true);
+  if (!accTyNameOpt)
+    return failure();
+  std::string accVarName = accName.str();
+  std::string accAccTyName = *accTyNameOpt;
+  if (accTyName != accAccTyName) {
+    accVarName = accVarName + "_acc";
+    os << accAccTyName << " " << accVarName << " = " << accAccTyName << "("
+       << accName << ");\n";
+  }
+
+  auto lhsShape = cast<VectorType>(lhs.getType()).getShape();
+  auto rhsShape = cast<VectorType>(rhs.getType()).getShape();
+  os << accVarName << " = mac_" << lhsShape[0] << "x" << lhsShape[1] << "_"
+     << rhsShape[0] << "x" << rhsShape[1] << "(";
+  os << lhsName << ", " << rhsName << ", " << accVarName << ")";
+
+  if (accTyName != accAccTyName)
+    os << ";\n" << accName << " = " << accTyName << "(" << accVarName << ")";
```

Comment:
I was hoping that we can have something like below. Since this only happens for the bfloat16 case, I'm ok to leave it as-is.
```
%0 = aievec.cast %C {isResAcc = true} : vector<4x4xf32>, vector<4x4xf32>
%1 = aievec.matmul %A, %B, %0 : vector<4x8xbf16>, vector<8x4xbf16> into vector<4x4xf32>
%2 = aievec.cast %1 {isResAcc = false} : vector<4x4xf32>, vector<4x4xf32>
```